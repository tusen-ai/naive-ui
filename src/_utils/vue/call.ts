type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
