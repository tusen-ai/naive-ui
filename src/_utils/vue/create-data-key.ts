export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
