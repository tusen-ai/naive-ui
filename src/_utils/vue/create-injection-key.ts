import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
