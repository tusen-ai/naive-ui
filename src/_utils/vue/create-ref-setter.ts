import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
