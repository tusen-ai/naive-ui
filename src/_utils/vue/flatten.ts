import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
