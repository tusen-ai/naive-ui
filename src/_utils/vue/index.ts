export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
