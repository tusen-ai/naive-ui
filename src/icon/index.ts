export { iconProps, NIcon } from './src/Icon'
export type { IconProps } from './src/Icon'
