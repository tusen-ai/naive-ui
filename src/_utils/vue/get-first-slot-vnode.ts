import { type Slots, type VNode } from 'vue'
import { flatten } from './flatten'
import { warn } from '../naive'

export function getFirstSlotVNode (
  slots: Slots,
  slotName = 'default',
  props: unknown = undefined
): VNode | null {
  const slot = slots[slotName]
  if (!slot) {
    warn('getFirstSlotVNode', `slot[${slotName}] is empty`)
    return null
  }
  const slotContent = flatten(slot(props))
  // vue will normalize the slot, so slot must be an array
  if (slotContent.length === 1) {
    return slotContent[0]
  } else {
    warn('getFirstSlotVNode', `slot[${slotName}] should have exactly one child`)
    return null
  }
}
