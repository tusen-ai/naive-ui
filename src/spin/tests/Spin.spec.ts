import { h } from 'vue'
import { mount } from '@vue/test-utils'
import { NSpin } from '../index'
import { Reload } from '@vicons/ionicons5'
import { NIcon } from '../../icon'

describe('n-spin', () => {
  it('should work with import on demand', () => {
    mount(NSpin)
  })

  it('should work with `show` prop', async () => {
    const wrapper = mount(NSpin, {
      props: {
        show: true
      },
      slots: {
        default: () => 'test'
      }
    })
    expect(wrapper.find('.n-spin-content').classes()).toContain(
      'n-spin-content--spinning'
    )
    wrapper.unmount()
  })

  it('should work with icon slot', () => {
    const wrapper = mount(NSpin, {
      slots: {
        icon: () =>
          h(NIcon, null, {
            default: () => h(Reload)
          })
      }
    })

    expect(wrapper.findComponent(NIcon).exists()).toBe(true)
    expect(wrapper.findComponent(Reload).exists()).toBe(true)
    wrapper.unmount()
  })

  it('rotate should work on icon slot', async () => {
    const wrapper = mount(NSpin, {
      slots: {
        icon: () =>
          h(NIcon, null, {
            default: () => h(Reload)
          })
      }
    })
    expect(wrapper.find('.n-spin--rotate').exists()).toBe(true)
    await wrapper.setProps({
      rotate: false
    })
    expect(wrapper.find('.n-spin--rotate').exists()).toBe(false)
    wrapper.unmount()
  })

  it('should work with `size` prop', async () => {
    ;(['small', 'medium', 'large', 71] as const).forEach((item) => {
      const wrapper = mount(NSpin, {
        props: {
          size: item
        }
      })
      expect(wrapper.find('.n-spin').attributes('style')).toMatchSnapshot()
      wrapper.unmount()
    })
  })

  it('should work with `default` slot', async () => {
    const wrapper = mount(NSpin, {
      slots: {
        default: () => 'test'
      }
    })
    expect(wrapper.find('.n-spin-container').exists()).toBe(true)
    expect(wrapper.find('.n-spin-content').text()).toBe('test')
    wrapper.unmount()
  })

  it('should work with `strokeWidth` prop', () => {
    const wrapper = mount(NSpin, {
      props: {
        strokeWidth: 40,
        size: 'medium'
      }
    })

    expect(wrapper.find('circle').attributes('stroke-width')).toEqual('40')
    wrapper.unmount()
  })
})
