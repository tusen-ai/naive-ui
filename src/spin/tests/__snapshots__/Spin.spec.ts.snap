// Vitest Snapshot v1

exports[`n-spin > should work with \`size\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 28px; --n-color: #18a058; --n-text-color: #18a058;"`;

exports[`n-spin > should work with \`size\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 34px; --n-color: #18a058; --n-text-color: #18a058;"`;

exports[`n-spin > should work with \`size\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 40px; --n-color: #18a058; --n-text-color: #18a058;"`;

exports[`n-spin > should work with \`size\` prop 4`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 71px; --n-color: #18a058; --n-text-color: #18a058;"`;

exports[`n-spin should work with \`size\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 28px; --n-color: #18a058; --n-text-color: #18a058;"`;

exports[`n-spin should work with \`size\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 34px; --n-color: #18a058; --n-text-color: #18a058;"`;

exports[`n-spin should work with \`size\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 40px; --n-color: #18a058; --n-text-color: #18a058;"`;

exports[`n-spin should work with \`size\` prop 4`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-opacity-spinning: 0.5; --n-size: 71px; --n-color: #18a058; --n-text-color: #18a058;"`;
